Centralize gRPC error mapping

npepinpe opened the pull request:
## Description

This PR centralizes gRPC error mapping by introducing a new `ErrorMappingStreamObserver`. The new observer will decorate an existing one, and `onError`, map the given exception to a `StatusRuntimeException` (assuming it isn't one already). It will also suppress the cancel exception if the decorated observer is a `ServerCallStreamObserver` (which it should be, as this is meant to be used only on the server side).

Part of the PR also introduces a new, richer error model as described by Google (see [here](https://cloud.google.com/apis/design/errors#error_model)), by first mapping errors to `com.google.rpc.Status`, before converting it to exceptions via `StatusProto`. Later on, we can then add details to errors, or aggregate multiple errors into composite types (see for example these [types](https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto) which we could use), but this was not done as part of the scope of this PR.

## Related issues

closes #5237 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/grpc/ErrorMappingStreamObserver.java:
```diff
@@ -0,0 +1,130 @@
+ * A {@link StreamObserver} decorator which will map errors to {@link StatusException} before
+ * passing them on to the {@code delegate}.
+ *
+ * <p>Additionally, if the decorated {@link StreamObserver} is a {@link ServerCallStreamObserver},
```
Actually you only accept ServerCallStreamObserver

npepinpe replied:
True, I wrote this before I changed the interface 😅

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/grpc/GrpcErrorMapper.java:
```diff
@@ -0,0 +1,139 @@
+      // When there is back pressure, there will be a lot of `RESOURCE_EXHAUSTED` errors and the log
+      // can get flooded, so log them at the lowest level possible
+      if (status.getCode() != Code.RESOURCE_EXHAUSTED.getNumber()) {
+        logger.error("Expected to handle gRPC request, but received error from broker", error);
```
Currently thinking whether it really necessary to error in gateway in general, since they properly logged in the broker already and when the client receives the error

npepinpe replied:
I'm not sure we log all errors in the broker. For example, partition mismatch i think is not logged, maybe others?

Zelldon replied:
But is this really an error then is the question? It is kind of expected that this can happen - ideally we should just retry that in this case

npepinpe replied:
My rule of thumb in general is, assuming a client-server setup, if the server returns an error, the client should log it, not the server. However, the server might log error details that are not returned to the client - i.e. sensitive info, information that isn't helpful for the client, etc. 

In this particular case, we return an INTERNAL error to the client (e.g. 500 - which, by the way, is probably the wrong code to return for `PARTITION_LEADER_MISMATCH` :sweat_smile:), with very little information from the client point of view. So we need to log it in the broker/gateway. From the broker point of view though, you're right, this is an expected error - hence why I'd log it on the "client" side, i.e. the gateway in this instance. But like I said, bad example, as `PARTITION_LEADER_MISMATCH` should probably return `UNAVAILABLE` or something like this, indicating the user can retry the request :sweat_smile:

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/impl/broker/RequestRetryHandler.java:
```diff
@@ -27,6 +27,7 @@
  * all partitions are tried. The request is retried only for specific errors such as connection
  * errors or resource exhausted errors. The request is not retried for time outs.
  */
+@SuppressWarnings("java:S119")
```
:sweat_smile:

npepinpe replied:
SonarCloud complains that the type names don't conform to the format. I don't think we ever decided on one, but I suppose we could just change the format on sonarcloud and not have to suppress those warnings

Zelldon replied:
What type names we are talking here? That interface should be an adjective?

Zelldon commented on gateway/src/test/java/io/zeebe/gateway/grpc/GrpcErrorMapperTest.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Test
+  void shouldTraceLogResourceExhaustedError() {
```
```suggestion
  void shouldLogResourceExhaustedErrorOnTrace() {
```

Zelldon reviewed (commented):
Thanks @npepinpe just some small comments

npepinpe commented on gateway/src/main/java/io/zeebe/gateway/EndpointManager.java:
```diff
@@ -119,7 +111,8 @@ private void addPartitionInfoToBrokerInfo(
       final ActivateJobsRequest request,
       final StreamObserver<ActivateJobsResponse> responseObserver) {
-    activateJobsHandler.activateJobs(request, responseObserver);
+    activateJobsHandler.activateJobs(
```
Wasn't aware of it, but its a good idea

npepinpe commented:
I applied your review feedback, adding `EndpointWrapper` as `GatewayGrpcService`, and a new interface `ServerStreamObserver`.

The idea of `GatewayGrpcService` is to start moving in a direction where we decouple the gRPC entry point from the actual gateway logic. Moving forward, we could move the request mapping/response mapping to the gRPC specific stuff, and have the endpoint handlers already use the mapped requests/response so we they can focus on the business logic and not the gRPC specific stuff.

The `ServerStreamObserver` I added - I get that it seems like additional fuzz (and it might be, maybe I'm overthinking this), but the idea is that `ServerCallStreamObserver` is not only a concrete class, but it has a wider API (so more difficult to mock), and is an experimental API. So I'd rather have our own small interface, and we simply need to make sure we're mapping the implementations properly instead of polluting all the call sites with the experimental one.

I also added missing branches for broker errors/rejections; the main difference is now mapping `PROCESSING_ERROR` to `INTERNAL` (wasn't sure about that one though, happy to discuss), and `PARTITION_LEADER_MISMATCH` to `UNAVAILABLE` (since it most likely is transient, and retrying should fix it).

Let me know what you think :+1:

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/EndpointManager.java:
```diff
@@ -107,48 +106,42 @@ private void addPartitionInfoToBrokerInfo(
-  @Override
   public void completeJob(
       final CompleteJobRequest request,
       final StreamObserver<CompleteJobResponse> responseObserver) {
```
Why are some methods are changed to use the new interface and some not?

npepinpe replied:
Only the activateJobsHandler really needs the new type actually...but it makes sense to change all methods to accept it.

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/grpc/GrpcErrorMapper.java:
```diff
@@ -99,6 +99,20 @@ private Status mapBrokerErrorToStatus(final BrokerError error) {
+      case NULL_VAL:
       default:
         builder.setCode(Code.INTERNAL_VALUE);
         message =
```
Maybe it makes more sense to return the string representation of the error code,  instead of returning the code in the error message? Then it is more clear for the user/client?

npepinpe replied:
Probably, - I can't remember now why we return the code and not the string representation.

npepinpe replied:
By string representation, you mean of the code (in this case that's what we do), or you mean the error message? The idea why we don't put the error message is that for these errors, we've logged it on the broker side, and it's probably not very useful for the client (also, it may contain sensitive info for some customers such as file path, system info, etc.)

Zelldon replied:
Ok my fault sorry. I thought this returns the int representation.

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/grpc/ServerStreamObserver.java:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @param <GrpcResponseT> the expected gRPC response type
+ */
+@SuppressWarnings("java:S119")
```
It is a bit weird with the suppresswarnings to be honest

npepinpe replied:
Like I said, it's because sonar cloud is set up to warn you if you don't use all caps for your types. We can also change the SonarCloud config instead :+1:

Zelldon reviewed (changes requested):
I like it :+1: I had just some smaller comments

npepinpe commented:
Applied review feedback :+1:

Zelldon reviewed (approved):
Thanks @npepinpe good job :+1:

npepinpe commented:
bors r+
